src/zenml/zen_server/utils.py

```diff
@@ -318,6 +335,70 @@ def decorated(*args: Any, **kwargs: Any) -> Any:
     return cast(F, decorated)
 
 
+def rate_limit_requests(
+    api_route: str,
+    day_limit: Optional[int] = None,
+    minute_limit: Optional[int] = None,
+) -> Callable[..., Any]:
+    """Decorator to handle exceptions in the API.
+
+    Args:
+        api_route: API route.
+        day_limit: Number of requests allowed per day.
+        minute_limit: Number of requests allowed per minute.
+
+    Returns:
+        Decorated function.
+
+    Raises:
+        ValueError: If both day_limit and minute_limit are None.
+    """
+    limiter = Limiter(
+        key_func=get_ipaddr, enabled=server_config().rate_limit_enabled
+    )
+    if day_limit is None and minute_limit is None:
+        raise ValueError("Pass either day or minuter limits, or both.")
+    limit_str = ""
+    if day_limit:
+        limit_str += f"{day_limit}/day;"
+    if minute_limit:
+        limit_str += f"{minute_limit}/minute"
+
+    def decorator(func: F) -> F:
+        @limiter.limit(limit_str)
+        @wraps(func)
+        def decorated(
+            request: Request,
+            response: Response,
+            auth_form_data: "OAuthLoginRequestForm",
+        ) -> Any:
+            try:
+                ret = func(
+                    request=request,
+                    response=response,
+                    auth_form_data=auth_form_data,
+                )
+            except Exception as e:
+                raise e
+            else:
+                if server_config().rate_limit_enabled:
+                    identifier = limiter._key_func(request)
+                    if "memory" not in (limiter._storage.STORAGE_SCHEME or []):
+                        raise NotImplementedError(
+                            "Limiter not running in memory is not yet supported."
+                        )
+                    for k in limiter._storage.storage.keys():  # type: ignore[union-attr]
+                        if api_route in k:
+                            _, identifier_ = k.split("//")[0].split("/")
+                            if identifier_ == identifier:
+                                limiter._storage.storage[k] -= 1  # type: ignore[union-attr]
```

Comment:
You're digging deep into the slowapi implementation, essentially hacking its internal state, to achieve a behavior that the tool clearly wasn't designed for. This makes it really difficult to maintain this, for the following reasons:

* it's really difficult to understand what this hack does without knowing a lot more about the slowapi internals
* the slowapi library itself is still in alpha stage, which means it's still subject to change a lot over the next releases and that means it's very probable that this code will be invalidated when we decide to update to a newer version

Instead, we should try to use the public (documented) APIs to achieve our goal, or work to implement these new features in the original project. As a last resort, this could be achieved by temporarily sub-classing the `Limiter` class and adding the missing functionality there, if this is even possible.

Reply:
I gave this more thought and spend a lot of time trying to find a solution for this that doesn't require us to mess with the Limiter's internals and I realized we don't even have a good understanding of what it is we want to achieve, so I'll try to summarize it here, maybe get some alignment.

IMO there are two end goals:

1. make it really difficult for someone to use brute-force to hack into an account (e.g. guess a username, password, an API token or invitation token) by allowing only a limited number of failed login attempts. We should also make sure that even if a successful login attempt is made after a series of failed ones, nothing in the server's response can be used to infer this. 
2. don't impact "good" users that use valid credentials to log in on the first try. Even if they fail to login for a couple of times, we should make it easy for them to recover.

I think the following mechanism achieves both goals, but I'm open to suggestions for how we can make it more robust:

1. the limiter should be applied for a source IP (i.e. start counting requests and raising limiter errors) only when the first failed login happens, but not before that.
2. the limiter should stop being applied for a source IP until the following happens, at which time, the limiter should be reset to its original state (i.e. its counters should be reset):
  * a successful login request is received
  * there is no time restriction in place (i.e. the request would be allowed by the limiter)

If we can't achieve this with the official slowapi features, I'd be more inclined to implement our own, slim, in-memory limiter, targeted to our reduced set of features, rather than hacking slowapi.